encoding/ccf/encode.go

```diff
@@ -0,0 +1,1900 @@
+	defer func() {
+		// Recover panic error if there is any.
+		if r := recover(); r != nil {
+			// Don't recover Go errors, internal errors, or non-errors.
+			switch r := r.(type) {
+			case goRuntime.Error, cadenceErrors.InternalError:
+				panic(r)
+			case error:
+				err = r
+			default:
+				panic(r)
+			}
+		}
+
+		// Add context to error if there is any.
+		if err != nil {
+			err = fmt.Errorf(
+				"ccf: failed to encode value (type %T, %q): %s",
+				value,
+				value.Type().ID(),
+				err,
+			)
+		}
+	}()
+
+	// Traverse value to find all composite types.
+	types, tids := compositeTypesFromValue(value)
+
+	if len(types) == 0 {
+		// Encode top level message: ccf-type-and-value-message.
+		err = e.encodeTypeAndValue(value, tids)
+	} else {
+		// Encode top level message: ccf-typedef-and-value-message.
+		err = e.encodeTypeDefAndValue(value, types, tids)
+	}
+	if err != nil {
+		return err
+	}
+
+	return e.enc.Flush()
+}
+
+// encodeTypeDefAndValue encodes type definition and value as
+// language=CDDL
+// ccf-typedef-and-value-message =
+//
+//	; cbor-tag-typedef-and-value
+//	#6.129([
+//	  typedef: composite-typedef,
+//	  type-and-value: inline-type-and-value
+//	])
+func (e *Encoder) encodeTypeDefAndValue(
+	value cadence.Value,
+	types []cadence.Type,
+	tids ccfTypeIDByCadenceType,
+) error {
+	// Encode tag number cbor-tag-typedef-and-value and array head of length 2.
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, CBORTagTypeDefAndValue,
+		// array, 2 items follow
+		0x82,
+	})
+	if err != nil {
+		return err
+	}
+
+	// element 0: typedef
+	err = e.encodeTypeDefs(types, tids)
+	if err != nil {
+		return err
+	}
+
+	// element 1: type and value
+	return e.encodeInlineTypeAndValue(value, tids)
+}
+
+// encodeTypeAndValue encodes type and value as
+// language=CDDL
+// ccf-type-and-value-message =
+//
+//	; cbor-tag-type-and-value
+//	#6.130(inline-type-and-value)
+func (e *Encoder) encodeTypeAndValue(value cadence.Value, tids ccfTypeIDByCadenceType) error {
+	// Encode tag number
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, CBORTagTypeAndValue,
+	})
+	if err != nil {
+		return err
+	}
+
+	return e.encodeInlineTypeAndValue(value, tids)
+}
+
+// encodeTypeAndValueWithNoTag encodes inline type and value as
+// language=CDDL
+// inline-type-and-value = [
+//
+//	type: inline-type,
+//	value: value,
+//
+// ]
+func (e *Encoder) encodeInlineTypeAndValue(value cadence.Value, tids ccfTypeIDByCadenceType) error {
+	// Encode array head of length 2.
+	err := e.enc.EncodeRawBytes([]byte{
+		// array, 2 items follow
+		0x82,
+	})
+	if err != nil {
+		return err
+	}
+
+	runtimeType := value.Type()
+
+	// element 0: inline-type
+	err = e.encodeInlineType(runtimeType, tids)
+	if err != nil {
+		return err
+	}
+
+	// element 1: value
+	return e.encodeValue(value, runtimeType, tids)
+}
+
+// encodeTypeDefs encodes composite/interface type definitions as
+// language=CDDL
+// composite-typedef = [
+//
+//	+(
```

Comment:
```suggestion
//  ; one-or-more instead of zero-or-more because:
//  ; - when encoding a primitive type, such as boolean or string, `ccf-type-and-value-message` is used (no `composite-typedef` at all)
//  ; - when encoding a composite type, such as event, `ccf-typedef-and-value-message` is used, which encodes at least one `composite-typedef`
//	+ (
```